// avenger/lib/DSL.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.observe = exports.param = void 0;
var Query_1 = require("./Query");
var TaskEither_1 = require("fp-ts/lib/TaskEither");
var Strategy_1 = require("./Strategy");
function param() {
    return Query_1.queryStrict(function (a) { return TaskEither_1.taskEither.of(a); }, Strategy_1.refetch);
}
exports.param = param;
var Query_2 = require("./Query");
Object.defineProperty(exports, "queryShallow", { enumerable: true, get: function () { return Query_2.queryShallow; } });
Object.defineProperty(exports, "queryStrict", { enumerable: true, get: function () { return Query_2.queryStrict; } });
var Strategy_2 = require("./Strategy");
Object.defineProperty(exports, "available", { enumerable: true, get: function () { return Strategy_2.available; } });
Object.defineProperty(exports, "refetch", { enumerable: true, get: function () { return Strategy_2.refetch; } });
Object.defineProperty(exports, "expire", { enumerable: true, get: function () { return Strategy_2.expire; } });
var Query_3 = require("./Query");
Object.defineProperty(exports, "compose", { enumerable: true, get: function () { return Query_3.compose; } });
Object.defineProperty(exports, "product", { enumerable: true, get: function () { return Query_3.product; } });
var observe_1 = require("./observe");
exports.observe = observe_1.observeShallow;
var invalidate_1 = require("./invalidate");
Object.defineProperty(exports, "invalidate", { enumerable: true, get: function () { return invalidate_1.invalidate; } });
var command_1 = require("./command");
Object.defineProperty(exports, "command", { enumerable: true, get: function () { return command_1.command; } });
